### PR TITLE
passfd-io: fix FIFO opening and vsock handling

### PR DESCRIPTION
**Pull request by Tim-Zhang**
- dragonball: fix EPOLLHUP/EPOLLERR events handling in vsock
- agent: Fix close_stdin for passfd-io
- runtime-rs: fix FIFO handling

Fixes: #9334

#### 1. Start a container (using terminal) -- OK

Run a normal container with `-t` and a simple command.
- [x] The console should be responsive.
- [x] The command should work normally.

<details>

```sh
$ sudo ctr run --runtime io.containerd.run.kata.v2 --rm -t docker.io/library/busybox:latest hello id
uid=0(root) gid=0(root) groups=0(root),10(wheel)

$ sudo ctr run --runtime io.containerd.run.kata.v2 --rm -t docker.io/library/ubuntu:latest hello id
uid=0(root) gid=0(root) groups=0(root)

$ sudo ctr run --runtime io.containerd.run.kata.v2 --rm -t docker.io/library/ubuntu:latest hello sh
# id
uid=0(root) gid=0(root) groups=0(root)
# exit
```

</details>

#### 2. Start a container (not using terminal) -- OK

Run a normal container with a simple command, not using a terminal.
- [x] The console should be responsive.
- [x] The command should work normally.

<details>

```sh
$ sudo ctr run --runtime io.containerd.run.kata.v2 --rm docker.io/library/busybox:latest hello id
uid=0(root) gid=0(root) groups=0(root),10(wheel)

$ sudo ctr run --runtime io.containerd.run.kata.v2 --rm docker.io/library/ubuntu:latest hello id
uid=0(root) gid=0(root) groups=0(root)

$ sudo ctr run --runtime io.containerd.run.kata.v2 --rm docker.io/library/ubuntu:latest hello sh
id
uid=0(root) gid=0(root) groups=0(root)
exit
```

</details>

#### 3. Start a detached container (using terminal) -- OK

Run a detached container, with a terminal.
- [x] The initial task status should be 'RUNNING'.
- [x] CPU usage shouldn't be 100%.

<details>

```sh
$ sudo ctr run --runtime io.containerd.run.kata.v2 -d -t docker.io/library/busybox:latest hello

$ sudo ctr container ls
CONTAINER    IMAGE                               RUNTIME
hello        docker.io/library/busybox:latest    io.containerd.run.kata.v2

$ sudo ctr task ls
TASK     PID        STATUS
hello    (pid number)    RUNNING
```

</details>

#### 4. Reattach to a detached container (using terminal) -- OK

Run a detached container, with a terminal.
Attach to the container.
Kill the `ctr` process to detach.
Attach to the container again.
- [x] Can be attached and re-attached.
- [x] The attached terminal should be responsive.

<details>

```sh
$ sudo ctr run --runtime io.containerd.run.kata.v2 -d -t docker.io/library/busybox:latest hello

$ sudo ctr t attach hello
/ #
/ # id
uid=0(root) gid=0(root) groups=0(root),10(wheel)
/ #
```

Now in another session:
```sh
$ sudo pkill -9 -f ctr
```

Re-attach to the same container:

```sh
Killed
$ sudo ctr t attach hello

/ # id
uid=0(root) gid=0(root) groups=0(root),10(wheel)
/ #

```

</details>

#### 5. Start a detached container (not using terminal) -- OK

Run a detached container, not using a terminal.
- [x] The container status should be 'RUNNING'.
- [x] CPU usage shouldn't be 100%.

<details>

```sh
$ sudo ctr run --runtime io.containerd.run.kata.v2 -d docker.io/library/busybox:latest hello

$ sudo ctr container ls
CONTAINER    IMAGE                               RUNTIME
hello        docker.io/library/busybox:latest    io.containerd.run.kata.v2

$ sudo ctr task ls
TASK     PID        STATUS
hello    (pid number)    RUNNING
```

</details>

#### 6. Reattach to a detached container (not using terminal) -- OK

Run a detached container, not using a terminal.
Attach to the container.
Kill the `ctr` process to detach.
Attach to the container again.
- [x] The container should be able to be attached and re-attached.
- [x] The attached terminal should be responsive.

<details>

```sh
$ sudo ctr run --runtime io.containerd.run.kata.v2 -d docker.io/library/busybox:latest hello

$ sudo ctr t attach hello
echo Hello
Hello
```

Now in another session:
```sh
$ sudo pkill -9 -f ctr
```

Re-attach to the same container:
```sh
Killed
$ sudo ctr t attach hello
echo Hello again
Hello again

```

</details>

#### 7. Execute command (using terminal) in an existing container -- OK

Run a container. Execute simple commands in the container using a terminal.
- [x] Can execute commands in the container.
- [x] The console should be responsive.

<details>

```sh
$ sudo ctr run --runtime io.containerd.run.kata.v2 -d docker.io/library/busybox:latest hello

$ sudo ctr t exec -t --exec-id foobar hello id
uid=0(root) gid=0(root) groups=0(root),10(wheel)

$ sudo ctr t exec -t --exec-id foobar hello sh
/ # id
uid=0(root) gid=0(root) groups=0(root),10(wheel)
/ # exit
```

</details>

#### 8. Execute command (not using terminal) in an existing container -- OK

Run a container. Execute simple commands in the container, not using a terminal.
- [x] Can execute commands in the container.
- [x] The console should be responsive.

<details>

```sh
$ sudo ctr run --runtime io.containerd.run.kata.v2 -d docker.io/library/busybox:latest hello

$ sudo ctr t exec --exec-id foobar hello id
uid=0(root) gid=0(root) groups=0(root),10(wheel)

$ sudo ctr t exec --exec-id foobar hello sh
id
uid=0(root) gid=0(root) groups=0(root),10(wheel)
exit
```

</details>

#### 9. Execute command (not using terminal, pipe stdin) in an existing container -- OK

Run a container. Execute simple commands in the container with pipe as stdin.
- [x] Can execute commands in the container.
- [x] The IO data should not changed.

<details>

```sh
$ sudo ctr run --runtime io.containerd.run.kata.v2 -d docker.io/library/busybox:latest hello

# Word count
$ printf "aaa\nbbb\nccc\n" | sudo ctr t exec --exec-id foobar hello wc
        3         3        12

# Large file count
$ head -c 1000000000 /dev/random > /tmp/input
$ cat /tmp/input | wc -c
1000000000
$ cat /tmp/input | sudo ctr t exec --exec-id foobar hello wc -c
1000000000
$ cat /tmp/input | sudo ctr t exec --exec-id foobar hello cat | wc -c
1000000000

# Large file copy
$ head -c 1000000000 /dev/random > /tmp/input
$ cat /tmp/input | sudo ctr t exec --exec-id foobar hello cat > /tmp/output
$ diff -q /tmp/input /tmp/output
```

</details>

### CODE REVIEW

**Review by Apokleos — APPROVED** [review]
Thx @Tim-Zhang LGTM!

File: src/agent/rustjail/src/container.rs [hunk 1]
```diff
@@ -1002,38 +1002,24 @@ impl BaseContainer for LinuxContainer {
...
-                                            debug!(logger, "copy from stdin to term_master end: {:?}", res);
-                                            break;
-                                        }
-                                        Ok(n) => {
-                                            if term_master.write_all(&buf[..n]).await.is_err() {
-                                                break;
-                                            }
-                                        }
-                                    }
+                            let res = stdin_stream.read(&mut buf).await;
```
Comment (lifupan) [inline src/agent/rustjail/src/container.rs:1010]: Here can we  use the tokio io copy directly?

File: src/agent/rustjail/src/container.rs [hunk 2]
```diff
@@ -1070,37 +1056,23 @@ impl BaseContainer for LinuxContainer {
...
-                                            break;
-                                        }
-                                        Ok(n) => {
-                                            if parent_stdin.write_all(&buf[..n]).await.is_err() {
-                                                break;
-                                            }
-                                        }
-                                    }
+                            let res = stdin_stream.read(&mut buf).await;
+                            match res {
```
Comment (lifupan) [inline src/agent/rustjail/src/container.rs:1064]: The same here.

File: src/agent/rustjail/src/container.rs (hunk 1 above)
Comment (Tim-Zhang) [reply src/agent/rustjail/src/container.rs]: updated, thanks

File: src/agent/rustjail/src/container.rs (hunk 2 above)
Comment (Tim-Zhang) [reply src/agent/rustjail/src/container.rs]: done

**Review by lifupan — COMMENTED** [review]
LGTM

**Review by lifupan — APPROVED** [review]

**Pull request merged**